optimizely/helpers/condition.py

```diff
@@ -12,115 +12,348 @@
+  def less_than_evaluator(self, index):
+    """ Evaluate the given less than match condition for the user attributes
+
+    Args:
+      index: Index of the condition to be evaluated
+
+    Returns:
+      Boolean:
+        - True if the user attribute value is less than the condition value
+        - False if the user attribute value is greater than or equal to the condition value
+      None: if the condition value isn't finite or the user attribute value isn't finite
+    """
+    condition_value = self.condition_data[index][1]
+    user_value = self.attributes.get(self.condition_data[index][0])
+
+    if not self.is_finite(condition_value) or not self.is_finite(user_value):
+      return None
+
+    return user_value < condition_value
+
+  def substring_evaluator(self, index):
+    """ Evaluate the given substring match condition for the given user attributes
+
+    Args:
+      index: Index of the condition to be evaluated
+
+    Returns:
+      Boolean:
+        - True if the condition value is a substring of the user attribute value
+        - False if the condition value is not a substring of the user attribute value
+      None: if the condition value isn't a string or the user attribute value isn't a string
+    """
+    condition_value = self.condition_data[index][1]
+    user_value = self.attributes.get(self.condition_data[index][0])
+
+    if not isinstance(condition_value, string_types) or not isinstance(user_value, string_types):
+      return None
+
+    return condition_value in user_value
+
+  EVALUATORS_BY_MATCH_TYPE = {
+    ConditionMatchTypes.EXACT: exact_evaluator,
+    ConditionMatchTypes.EXISTS: exists_evaluator,
+    ConditionMatchTypes.GREATER_THAN: greater_than_evaluator,
+    ConditionMatchTypes.LESS_THAN: less_than_evaluator,
+    ConditionMatchTypes.SUBSTRING: substring_evaluator
   }
 
-  def evaluate(self, conditions):
-    """ Top level method to evaluate audience conditions.
+  def evaluate(self, index):
+    """ Given a custom attribute audience condition and user attributes, evaluate the
+        condition against the attributes.
 
     Args:
-      conditions: Nested list of and/or conditions.
-                  Ex: ['and', operand_1, ['or', operand_2, operand_3]]
+      index: Index of the condition to be evaluated
 
     Returns:
-      Boolean result of evaluating the conditions evaluate
+      Boolean:
+        - True if the user attributes match the given condition
+        - False if the user attributes don't match the given condition
+      None: if the user attributes and condition can't be evaluated
     """
 
-    if isinstance(conditions, list):
-      if conditions[0] in DEFAULT_OPERATOR_TYPES:
-        return self.OPERATORS[conditions[0]](self, conditions[1:])
-      else:
-        return False
+    if self.condition_data[index][2] != self.CUSTOM_ATTRIBUTE_CONDITION_TYPE:
+      return None
+
+    condition_match = self.condition_data[index][3]
+
+    if condition_match not in self.MATCH_TYPES:
```

Comment:
Nit: here, too, I wonder if we really need `MATCH_TYPES`. Something like
```python
if condition_match not in self.EVALUATORS_BY_MATCH_TYPE:
```
seems more natural given that we're using this to gate the expression
```
self.EVALUATORS_BY_MATCH_TYPE[condition_match]
```